### PR TITLE
EquipPlan, a class to describe what happens between TLS and HTTP

### PR DESCRIPTION
**Pull request by swankjesse**
After a successful TCP socket is acquired we may need a CONNECT tunnel
and/or TLS handshake. I'm using the word 'equip' to mean 'equipping a TCP
socket with state to carry HTTP' though I'm not attached to this name.

This is a step towards splitting the TCP and equip phases so that happy
eyeballs can do TCP in parallel but equipping in serial order.

### CODE REVIEW

File: okhttp/src/jvmTest/java/okhttp3/internal/connection/EquipPlanTest.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import java.security.cert.CertificateException
+import javax.net.ssl.SSLHandshakeException
+import javax.net.ssl.SSLSocket
+import okhttp3.ConnectionSpec
+import okhttp3.TlsVersion
+import okhttp3.tls.internal.TlsUtil.localhost
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class EquipPlanTest {
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/internal/connection/EquipPlanTest.kt:28]: Renamed from ConnectionSpecSelectorTest & converted to Kotlin.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ *  * TLS handshakes.
+ *
+ * We might need many attempts at each of these steps. Tunnels fail due to authentication
+ * challenges, and TLS handshakes fail due to mismatched protocol versions. When we need another
+ * attempt, this class tracks what to try next.
+ *
+ * Unlike routes, we don't know that we'll need another equip plan until the current one fails.
+ */
+internal data class EquipPlan(
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:45]: Note that this replaces ConnectionSpec, which did the same thing but statefully. (Also ConnectionSpec didn’t know about tunnels.)

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2022 Block, Inc.
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:2]: New corporation, who dis?

Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:2]: Should others change?

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  /** Returns the next plan to use after this, or null if no more compatible plans are available. */
+  fun nextConnectionSpec(
+    connectionSpecs: List<ConnectionSpec>,
+    sslSocket: SSLSocket
+  ): EquipPlan? {
+    for (i in connectionSpecIndex + 1 until connectionSpecs.size) {
+      if (connectionSpecs[i].isCompatible(sslSocket)) {
+        return copy(connectionSpecIndex = i)
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:75]: isTlsFallback = true?

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  val isTlsFallback: Boolean = false,
+) {
+
+  /** Returns this if its [connectionSpecIndex] is defined, or defines it otherwise. */
+  @Throws(IOException::class)
+  fun withInitialConnectionSpec(
+    connectionSpecs: List<ConnectionSpec>,
+    sslSocket: SSLSocket
+  ): EquipPlan {
+    if (connectionSpecIndex != -1) return this
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:58]: Not 100% clear on the flow here. Is this expected when connectionSpecIndex >= 0? Or is that unexpected and an error? I'll run through this in an IDE.

Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:58]: Never mind, the comment above is accurate and makes sense in the context of usage.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt [hunk 6]
```diff
@@ -199,81 +238,120 @@ class RealConnection(
...
+
+        // Create the wrapper over the connected socket.
+        val sslSocket = route.address.sslSocketFactory.createSocket(
+          rawSocket,
+          route.address.url.host,
+          route.address.url.port,
+          true /* autoClose */
+        ) as SSLSocket
+
+        val tlsEquipPlan = equipPlan.withInitialConnectionSpec(connectionSpecs, sslSocket)
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:310]: TIOLI: maybe withCurrentOrInitialConnectionSpec?

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt [hunk 7]
```diff
@@ -364,19 +460,10 @@ class RealConnection(
...
-    var sslSocket: SSLSocket? = null
     try {
-      // Create the wrapper over the connected socket.
-      sslSocket = sslSocketFactory!!.createSocket(
-        rawSocket, address.url.host, address.url.port, true /* autoClose */
-      ) as SSLSocket
-
-      // Configure the socket's ciphers, TLS versions, and extensions.
-      val connectionSpec = connectionSpecSelector.configureSecureSocket(sslSocket)
       if (connectionSpec.supportsTlsExtensions) {
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:467]: Tangential, should we remove this flag? Is it still useful in 2022, especially as most environments things like SNI are baked in and we don't control them, and don't actively disable them. Looking at configureTlsExtensions

Platform - does nothing.
Jdk9Platform - ALPN
Conscrypt - ALPN + setUseSessionTickets
AndroidSocketAdapter - ALPN + session tickets + SNI
Android10SocketAdapter - ALPN 
ConscryptSocketAdapter - ALPN + setUseSessionTickets

File: okhttp/src/jvmTest/java/okhttp3/internal/connection/EquipPlanTest.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    // MODERN_TLS is used here.
+    val attempt0 = equipPlan.withInitialConnectionSpec(connectionSpecs, socket)
+    connectionSpecs[attempt0.connectionSpecIndex].apply(socket, attempt0.isTlsFallback)
+    assertEnabledProtocols(socket, TlsVersion.TLS_1_2)
+    val attempt1 = attempt0.nextConnectionSpec(connectionSpecs, socket)
+    assertThat(attempt1).isNotNull()
+    socket.close()
+
+    // COMPATIBLE_TLS is used here.
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/internal/connection/EquipPlanTest.kt:71]: Nit: test for the fallback flag.

**Comment by yschimke** [comment]
This is looking really clean and understandable.

**Comment by yschimke** [comment]
```
[Unrecognized error] /home/runner/work/okhttp/okhttp/okhttp/build/classes/kotlin/jvm/main/okhttp3/internal/connection/EquipPlan.class: Undefined reference: int Integer.hashCode(int) 

> Task :okhttp:animalsnifferMain FAILED
```

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt:2]: Only for new files.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt (hunk 7 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:467]: I'm unsure and a bit risk averse with the things in Platform. Your call!

Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:467]: Let me think about it.  Likely yes.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt (hunk 4 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/EquipPlan.kt]: Ooooh good catch. Yes, fixed.

**Comment by swankjesse** [comment]
Thanks for catching the TLS fallback regression.

**Review by yschimke — APPROVED** [review]

**Comment by swankjesse** [comment]
Working towards https://github.com/square/okhttp/issues/506

**Pull request merged**